pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBuilderImpl.java

```diff
@@ -102,6 +101,11 @@ private ProducerBuilderImpl(PulsarClientImpl client, ProducerConfigurationData c
                     .failedFuture(new IllegalArgumentException("Topic name must be set on the producer builder"));
         }
 
+        if(conf.getCustomMessageRouter() != null
+                && conf.getMessageRoutingMode() != MessageRoutingMode.CustomPartition) {
```

Comment:
if a message router is set, but message routing mode is not `CustomPartition`, should not it be better to fail the builder to building a producer? because it means the users of pulsar are using the wrong settings or have a different expectation.

Reply:
Thanks @sijie for the review.
`messageRoutingMode` is set `RoundRobinPartition` as default so it is **optional**. In the light of this, we can not differentiate the following cases to decide if there is mismatch or not. So, both cases basically set as `RoundRobinPartition`.
```
.messageRouter(new AlwaysTwoMessageRouter())
```
and
```
.messageRoutingMode(MessageRoutingMode.RoundRobinPartition)
.messageRouter(new AlwaysTwoMessageRouter())
```
Also, `@deprecated ProducerConfiguration.setMessageRouter` is set `messageRoutingMode` **implicitly** for Custom Message Router (can be important to support `backward compatibility`)

Reply:
the part for setting `messageRoutingMode` to custom if a custom message router is set. however my point is in the case you showed, these two settings are conflicting with each other. I think the fix should be failing building producer if users have set conflicting settings.

```
.messageRoutingMode(MessageRoutingMode.RoundRobinPartition)
.messageRouter(new AlwaysTwoMessageRouter())
```

Reply:
I prefer giving the error as well, possibly in the `ProducerBuilder.build()` where we have the final picture.

Reply:
Thanks @sijie and @merlimat for feedbacks.

In general, i also prefer `fail-fast` as this kind of config mismatch.

My concern was the following case while `RoundRobinPartition` comes as **default** `MessageRoutingMode` through (`ProducerConfigurationData.messageRoutingMode = MessageRoutingMode.RoundRobinPartition;`)

```
.messageRouter(new AlwaysTwoMessageRouter())      // messageRoutingMode is not set by user. However, it is set as default at ProducerConfigurationData.messageRoutingMode
```
This case will also be matched / catched in same check logic (means: new proposed fail-fast check). 

I think this case can also fail-fast (Sample failure message: `when messageRouter is set, MessageRoutingMode should be set as CustomPartition`) in order to be set explicitly by user if makes sense.

Reply:
Should we leave `messageRoutingMode=null` in the builder and setting it to `RoundRobinPartition` only at the very end (if the messageRouter is not passed) ?

Reply:
Yep, exactly. i was thinking the same approach without breaking `messageRoutingMode`' s **optional** behaviour.
So, We can move this logic to final stage (`createAsync`) by removing default setting (`messageRoutingMode=null`).
I think all of us are on the same page so will be addressing in this direction. Thanks all again ;)